sgl-kernel/csrc/cpu/gemm_fp8.cpp

```diff
@@ -0,0 +1,356 @@
+            /*   A                  */ mat1 + mb_start * mat1_strideM,
+            /*   B                  */ mat2 + nb_start * K, // nb * BLOCK_N * K
+            /*   C                  */ out + mb_start * out_strideM + nb_start,
+            /*   Btmp               */ Btmp,
+            /*   Ctmp               */ Ctmp,
+            /*   scale              */ scale_ptr,
+            /*   bias               */ bias + nb_start,
+            /*   M                  */ mb_size,
+            /*   N                  */ nb_size,
+            /*   K                  */ K,
+            /*   lda                */ mat1_strideM,
+            /*   ldb                */ nb_size,
+            /*   ldc                */ out_strideM,
+            /*   brg                */ use_brgemm,
+            /*   blocks_k_per_group */ blocks_k_per_group,
+            /*   block_size_K       */ block_size_K);
+
+        // move to the next index
+        data_index_step(mb, MB, nb, NB);
+      }
+
+      if (use_brgemm) {
+        at::native::cpublas::brgemm_release();
+      }
+    });
+  });
+
+}
+
+} // anonymous namespace
+
+at::Tensor fp8_scaled_mm_cpu(at::Tensor& mat1, at::Tensor& mat2, at::Tensor& scales2,
+    std::vector<int64_t> block_size, std::optional<at::Tensor>& bias,
+    at::ScalarType out_dtype, bool is_vnni) {
+  RECORD_FUNCTION("sgl-kernel::fp8_scaled_mm_cpu", std::vector<c10::IValue>({mat1, mat2, scales2, block_size, bias}));
+
+  auto packed_w = is_vnni ? mat2 : convert_weight_packed(mat2);
+
+  CHECK_LAST_DIM_CONTIGUOUS_INPUT(mat1);
+  CHECK_INPUT(mat2);
+  CHECK_INPUT(scales2);
+  TORCH_CHECK(scales2.scalar_type() == at::kFloat,
+      "fp8_scaled_mm_cpu: expect scales2 to be float32.");
+
+  int64_t M = mat1.size(0);
+  int64_t N = mat2.size(0);
+  int64_t K = mat2.size(1);
+
+  CHECK_EQ(mat1.size(1), K);
+  CHECK_DIM(2, mat1);
+  CHECK_DIM(2, mat2);
+
+  TORCH_CHECK(block_size.size() == 2,
+      "fp8_scaled_mm_cpu: expect block_size.size() to be 2.");
+
+  int64_t block_size_N = block_size[0];
+  int64_t block_size_K = block_size[1];
+
+  constexpr int64_t BLOCK_N = block_size_n();
+  TORCH_CHECK(block_size_N >= BLOCK_N, "fp8_scaled_mm_cpu: expect block_size_N >= BLOCK_N");
+  TORCH_CHECK(block_size_K >= BLOCK_K, "fp8_scaled_mm_cpu: expect block_size_K >= BLOCK_K");
+  CHECK_EQ(scales2.size(0), div_up(N, block_size_N));
+  CHECK_EQ(scales2.size(1), div_up(K, block_size_K));
+
+  const auto st = mat1.scalar_type();
+  TORCH_CHECK(st == at::kBFloat16 || st == at::kHalf,
+      "fp8_scaled_mm_cpu: expect A to be bfloat16 or half.");
+  TORCH_CHECK(st == out_dtype,
+      "fp8_scaled_mm_cpu: expect A has same dtype with out_dtype.");
+  TORCH_CHECK(mat2.scalar_type() == at::kFloat8_e4m3fn,
+      "fp8_scaled_mm_cpu: expect mat2 to be fp8_e4m3.");
+  TORCH_CHECK(scales2.scalar_type() == at::kFloat,
+      "fp8_scaled_mm_cpu: expect scales to be float32.");
+  auto out = at::empty({M, N}, mat1.options().dtype(out_dtype));
+
```

Comment:
```suggestion
TORCH_CHECK(N % block_size_N == 0, "fp8_scaled_mm_cpu: expect N to be multiples of ", block_size_N);
TORCH_CHECK(K % block_size_K == 0, "fp8_scaled_mm_cpu: expect K to be multiples of ", block_size_K);

```

Reply:
In DeepSeek R1, we have the below layer which does not satisfy this check:

`model.layers.0.self_attn.kv_a_proj_with_mqa.weight`: N = 576, K = 7168 with block_size of N and K both equal to 128, `576 % 128 != 0` in this case.

The model satisfies the below conditation:
```cpp
CHECK_EQ(scales2.size(0), div_up(N, block_size_N));
CHECK_EQ(scales2.size(1), div_up(K, block_size_K));
```
which has been checked here: https://github.com/chunyuan-w/sglang/blob/chunyuan/pr_fp8_gemm_brgemm/sgl-kernel/csrc/cpu/gemm_fp8.cpp#L330-L331